### PR TITLE
Separate the interface for SBGEMMT from GEMMT

### LINKED ISSUE #4475: build fails on ppc64le (TARGET=POWER8) with GCC 14 due incompatible pointer types
Builds of release 0.3.26 fail on ppc64le (TARGET=POWER8) with GCC 14 due incompatible pointer types. This is on Fedora Rawhide. All other platforms in Fedora (aarch64, x86_64, s390x) build OK.
````
...
In file included from ../common.h:849,
                 from gemmt.c:37:
gemmt.c: In function ‘sbgemmt_’:
../common_stackalloc.h:59:10: error: assignment to ‘bfloat16 *’ {aka ‘short unsigned int *’} from incompatible pointer type ‘float *’ [-Wincompatible-pointer-types]
   59 |   BUFFER = stack_alloc_size ? stack_buffer : (TYPE *)blas_memory_alloc(1);
      |          ^
gemmt.c:481:25: note: in expansion of macro ‘STACK_ALLOC’
  481 |                         STACK_ALLOC(buffer_size, FLOAT, buffer);
      |                         ^~~~~~~~~~~
gemmt.c:504:70: error: passing argument 5 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  504 |                                 (gemv[(int)transa]) (j, k, 0, alpha, aa, lda,
      |                                                                      ^~
      |                                                                      |
      |                                                                      bfloat16 * {aka short unsigned int *}
gemmt.c:504:70: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:505:54: error: passing argument 7 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  505 |                                                      bb, incb, cc, 1, buffer);
      |                                                      ^~
      |                                                      |
      |                                                      bfloat16 * {aka short unsigned int *}
gemmt.c:505:54: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:505:71: error: passing argument 11 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  505 |                                                      bb, incb, cc, 1, buffer);
      |                                                                       ^~~~~~
      |                                                                       |
      |                                                                       bfloat16 * {aka short unsigned int *}
gemmt.c:505:71: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:507:70: error: passing argument 5 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  507 |                                 (gemv[(int)transa]) (k, j, 0, alpha, aa, lda,
      |                                                                      ^~
      |                                                                      |
      |                                                                      bfloat16 * {aka short unsigned int *}
gemmt.c:507:70: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:508:54: error: passing argument 7 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  508 |                                                      bb, incb, cc, 1, buffer);
      |                                                      ^~
      |                                                      |
      |                                                      bfloat16 * {aka short unsigned int *}
gemmt.c:508:54: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:508:71: error: passing argument 11 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  508 |                                                      bb, incb, cc, 1, buffer);
      |                                                                       ^~~~~~
      |                                                                       |
      |                                                                       bfloat16 * {aka short unsigned int *}
gemmt.c:508:71: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
../common_stackalloc.h:59:10: error: assignment to ‘bfloat16 *’ {aka ‘short unsigned int *’} from incompatible pointer type ‘float *’ [-Wincompatible-pointer-types]
   59 |   BUFFER = stack_alloc_size ? stack_buffer : (TYPE *)blas_memory_alloc(1);
      |          ^
gemmt.c:570:25: note: in expansion of macro ‘STACK_ALLOC’
  570 |                         STACK_ALLOC(buffer_size, FLOAT, buffer);
      |                         ^~~~~~~~~~~
gemmt.c:593:70: error: passing argument 5 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  593 |                                 (gemv[(int)transa]) (j, k, 0, alpha, a, lda, bb,
      |                                                                      ^
      |                                                                      |
      |                                                                      bfloat16 * {aka short unsigned int *}
gemmt.c:593:70: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:593:78: error: passing argument 7 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  593 |                                 (gemv[(int)transa]) (j, k, 0, alpha, a, lda, bb,
      |                                                                              ^~
      |                                                                              |
      |                                                                              bfloat16 * {aka short unsigned int *}
gemmt.c:593:78: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:594:67: error: passing argument 11 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  594 |                                                      incb, cc, 1, buffer);
      |                                                                   ^~~~~~
      |                                                                   |
      |                                                                   bfloat16 * {aka short unsigned int *}
gemmt.c:594:67: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:596:70: error: passing argument 5 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  596 |                                 (gemv[(int)transa]) (k, j, 0, alpha, a, lda, bb,
      |                                                                      ^
      |                                                                      |
      |                                                                      bfloat16 * {aka short unsigned int *}
gemmt.c:596:70: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:596:78: error: passing argument 7 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  596 |                                 (gemv[(int)transa]) (k, j, 0, alpha, a, lda, bb,
      |                                                                              ^~
      |                                                                              |
      |                                                                              bfloat16 * {aka short unsigned int *}
gemmt.c:596:78: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:597:67: error: passing argument 11 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  597 |                                                      incb, cc, 1, buffer);
      |                                                                   ^~~~~~
      |                                                                   |
      |                                                                   bfloat16 * {aka short unsigned int *}
gemmt.c:597:67: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
...
In file included from ../common.h:849,
                 from gemmt.c:37:
gemmt.c: In function ‘cblas_sbgemmt’:
../common_stackalloc.h:59:10: error: assignment to ‘bfloat16 *’ {aka ‘short unsigned int *’} from incompatible pointer type ‘float *’ [-Wincompatible-pointer-types]
   59 |   BUFFER = stack_alloc_size ? stack_buffer : (TYPE *)blas_memory_alloc(1);
      |          ^
gemmt.c:481:25: note: in expansion of macro ‘STACK_ALLOC’
  481 |                         STACK_ALLOC(buffer_size, FLOAT, buffer);
      |                         ^~~~~~~~~~~
gemmt.c:505:71: error: passing argument 11 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  505 |                                                      bb, incb, cc, 1, buffer);
      |                                                                       ^~~~~~
      |                                                                       |
      |                                                                       bfloat16 * {aka short unsigned int *}
gemmt.c:505:71: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:508:71: error: passing argument 11 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  508 |                                                      bb, incb, cc, 1, buffer);
      |                                                                       ^~~~~~
      |                                                                       |
      |                                                                       bfloat16 * {aka short unsigned int *}
gemmt.c:508:71: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
../common_stackalloc.h:59:10: error: assignment to ‘bfloat16 *’ {aka ‘short unsigned int *’} from incompatible pointer type ‘float *’ [-Wincompatible-pointer-types]
   59 |   BUFFER = stack_alloc_size ? stack_buffer : (TYPE *)blas_memory_alloc(1);
      |          ^
gemmt.c:570:25: note: in expansion of macro ‘STACK_ALLOC’
  570 |                         STACK_ALLOC(buffer_size, FLOAT, buffer);
      |                         ^~~~~~~~~~~
gemmt.c:594:67: error: passing argument 11 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  594 |                                                      incb, cc, 1, buffer);
      |                                                                   ^~~~~~
      |                                                                   |
      |                                                                   bfloat16 * {aka short unsigned int *}
gemmt.c:594:67: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
gemmt.c:597:67: error: passing argument 11 of ‘gemv[transa]’ from incompatible pointer type [-Wincompatible-pointer-types]
  597 |                                                      incb, cc, 1, buffer);
      |                                                                   ^~~~~~
      |                                                                   |
      |                                                                   bfloat16 * {aka short unsigned int *}
gemmt.c:597:67: note: expected ‘float *’ but argument is of type ‘bfloat16 *’ {aka ‘short unsigned int *’}
make[1]: *** [Makefile:1936: cblas_sbgemmt.o] Error 1
````

### PR DESCRIPTION
**Pull request by martin-frbg**
fixes #4475

### CODE REVIEW

File: test/compare_sgemm_sbgemm.c [hunk 1]
```diff
@@ -81,6 +81,16 @@ float16to32 (bfloat16_bits f16)
   return f32.v;
 }
 
+float
+float32to16 (float32_bits f32)
```
Comment (ChipKerchner) [inline test/compare_sgemm_sbgemm.c:85]: float32_bits is not defined in some cases.

```
compare_sgemm_sbgemm.c:121:39: error: passing 'float' to parameter of incompatible type 'float32_bits'
  121 |               AA[j * k + i].v = float32to16( A[j * k + i] );
      |                                              ^~~~~~~~~~~~
compare_sgemm_sbgemm.c:85:27: note: passing argument to parameter 'f32' here
   85 | float32to16 (float32_bits f32)
      |                           ^
compare_sgemm_sbgemm.c:122:39: error: passing 'float' to parameter of incompatible type 'float32_bits'
  122 |               BB[j * k + i].v = float32to16( B[j * k + i] );
      |                                              ^~~~~~~~~~~~
compare_sgemm_sbgemm.c:85:27: note: passing argument to parameter 'f32' here
   85 | float32to16 (float32_bits f32)
```

Comment (martin-frbg) [reply test/compare_sgemm_sbgemm.c:85]: strange, this is typedef'd as a union near the top of the file (same as with bfloat16_bits)

Comment (ChipKerchner) [reply test/compare_sgemm_sbgemm.c:85]: Sorry, it is defined.  It just is NOT compatible.  It believe you can NOT implicitly cast a float to a float32_bits union.  You need to explicitly cast in these cases (?)

Comment (ChipKerchner) [reply test/compare_sgemm_sbgemm.c:85]: Your changes also break some of the unit tests.

Comment (martin-frbg) [reply test/compare_sgemm_sbgemm.c:85]: which would that be ?

Comment (ChipKerchner) [reply test/compare_sgemm_sbgemm.c:85]: BTW, converting from float32 to bfloat16 is not as simple as shifting bits.  You need to take rounding into account (plus more if you are considering denormalization, etc)

Comment (ChipKerchner) [reply test/compare_sgemm_sbgemm.c:85]:
```
OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./test_sbgemm > SBBLAT3.SUMM
FATAL ERROR SBGEMM - Return code: 338349
```

Comment (martin-frbg) [reply test/compare_sgemm_sbgemm.c:85]: not sure why I did not see that... will fix in a followup PR

Comment (martin-frbg) [reply test/compare_sgemm_sbgemm.c:85]: #4491

**Pull request merged**